diff --git a/src/gui/kernel/qguiapplication.cpp b/src/gui/kernel/qguiapplication.cpp
index 86f4e5fdea..d542015c63 100644
--- a/src/gui/kernel/qguiapplication.cpp
+++ b/src/gui/kernel/qguiapplication.cpp
@@ -127,6 +127,7 @@ Q_CONSTINIT QGuiApplicationPrivate::QLastCursorPosition QGuiApplicationPrivate::
 
 Q_CONSTINIT QWindow *QGuiApplicationPrivate::currentMouseWindow = nullptr;
 
+Q_CONSTINIT QString QGuiApplicationPrivate::platformThemeOverride;
 Q_CONSTINIT QString QGuiApplicationPrivate::styleOverride;
 
 Q_CONSTINIT Qt::ApplicationState QGuiApplicationPrivate::applicationState = Qt::ApplicationInactive;
@@ -232,15 +233,6 @@ static void initThemeHints()
     touchDoubleTapDistance = QGuiApplicationPrivate::platformTheme()->themeHint(QPlatformTheme::TouchDoubleTapDistance).toInt();
 }
 
-static bool checkNeedPortalSupport()
-{
-#if QT_CONFIG(dbus)
-    return QFileInfo::exists("/.flatpak-info"_L1) || qEnvironmentVariableIsSet("SNAP");
-#else
-    return false;
-#endif // QT_CONFIG(dbus)
-}
-
 // Using aggregate initialization instead of ctor so we can have a POD global static
 #define Q_WINDOW_GEOMETRY_SPECIFICATION_INITIALIZER { Qt::TopLeftCorner, -1, -1, -1, -1 }
 
@@ -1221,12 +1213,11 @@ Q_LOGGING_CATEGORY(lcQpaPluginLoading, "qt.qpa.plugin");
 Q_LOGGING_CATEGORY(lcQpaTheme, "qt.qpa.theme");
 Q_LOGGING_CATEGORY(lcPtrDispatch, "qt.pointer.dispatch");
 
-static void init_platform(const QString &pluginNamesWithArguments, const QString &platformPluginPath, const QString &platformThemeName, int &argc, char **argv)
+static void init_platform(const QString &pluginNamesWithArguments, const QString &platformPluginPath, int &argc, char **argv)
 {
     qCDebug(lcQpaPluginLoading) << "init_platform called with"
         << "pluginNamesWithArguments" << pluginNamesWithArguments
-        << "platformPluginPath" << platformPluginPath
-        << "platformThemeName" << platformThemeName;
+        << "platformPluginPath" << platformPluginPath;
 
     QStringList plugins = pluginNamesWithArguments.split(u';', Qt::SkipEmptyParts);
     QStringList platformArguments;
@@ -1291,22 +1282,24 @@ static void init_platform(const QString &pluginNamesWithArguments, const QString
 
     // 1) Fetch the platform name from the environment if present.
     QStringList themeNames;
-    if (!platformThemeName.isEmpty()) {
-        qCDebug(lcQpaTheme) << "Adding" << platformThemeName << "from environment to list of theme names";
-        themeNames.append(platformThemeName);
+    if (!QGuiApplicationPrivate::platformThemeOverride.isEmpty()) {
+        qCDebug(lcQpaTheme) << "Adding" << QGuiApplicationPrivate::platformThemeOverride << "from environment to list of theme names";
+        themeNames.append(QGuiApplicationPrivate::platformThemeOverride);
     }
 
-    // 2) Special case - check whether it's a flatpak or snap app to use xdg-desktop-portal platform theme for portals support
-    if (checkNeedPortalSupport()) {
-        qCDebug(lcQpaTheme) << "Adding xdgdesktopportal to list of theme names";
-        themeNames.append(QStringLiteral("xdgdesktopportal"));
-    }
-
-    // 3) Ask the platform integration for a list of theme names
+    // 2) Ask the platform integration for a list of theme names
     const auto platformIntegrationThemeNames = QGuiApplicationPrivate::platform_integration->themeNames();
     qCDebug(lcQpaTheme) << "Adding platform integration's theme names to list of theme names:" << platformIntegrationThemeNames;
     themeNames += platformIntegrationThemeNames;
-    // 4) Look for a theme plugin.
+
+    // 3) Special case - prefer xdgdesktopportal
+    if (themeNames.indexOf(QStringLiteral("xdgdesktopportal")) > 0) {
+        qCDebug(lcQpaTheme) << "Moving xdgdesktopportal to the front of list of theme names";
+        themeNames.removeAll(QStringLiteral("xdgdesktopportal"));
+        themeNames.prepend(QStringLiteral("xdgdesktopportal"));
+    }
+
+    // 4) Look for a theme.
     for (const QString &themeName : std::as_const(themeNames)) {
         qCDebug(lcQpaTheme) << "Attempting to create platform theme" << themeName << "via QPlatformThemeFactory::create";
         QGuiApplicationPrivate::platform_theme = QPlatformThemeFactory::create(themeName, platformPluginPath);
@@ -1314,23 +1307,17 @@ static void init_platform(const QString &pluginNamesWithArguments, const QString
             qCDebug(lcQpaTheme) << "Successfully created platform theme" << themeName;
             break;
         }
-    }
 
-    // 5) If no theme plugin was found ask the platform integration to
-    // create a theme
-    if (!QGuiApplicationPrivate::platform_theme) {
-        for (const QString &themeName : std::as_const(themeNames)) {
-            qCDebug(lcQpaTheme) << "Attempting to create platform theme" << themeName << "via createPlatformTheme";
-            QGuiApplicationPrivate::platform_theme = QGuiApplicationPrivate::platform_integration->createPlatformTheme(themeName);
-            if (QGuiApplicationPrivate::platform_theme) {
-                qCDebug(lcQpaTheme) << "Successfully created platform theme" << themeName;
-                break;
-            }
+        qCDebug(lcQpaTheme) << "Attempting to create platform theme" << themeName << "via createPlatformTheme";
+        QGuiApplicationPrivate::platform_theme = QGuiApplicationPrivate::platform_integration->createPlatformTheme(themeName);
+        if (QGuiApplicationPrivate::platform_theme) {
+            qCDebug(lcQpaTheme) << "Successfully created platform theme" << themeName;
+            break;
         }
-        // No error message; not having a theme plugin is allowed.
     }
+    // No error message; not having a theme is allowed.
 
-    // 6) Fall back on the built-in "null" platform theme.
+    // 5) Fall back on the built-in "null" platform theme.
     if (!QGuiApplicationPrivate::platform_theme) {
         qCDebug(lcQpaTheme) << "Failed to create platform theme; using \"null\" platform theme";
         QGuiApplicationPrivate::platform_theme = new QPlatformTheme;
@@ -1479,7 +1466,7 @@ void QGuiApplicationPrivate::createPlatformIntegration()
         platformExplicitlySelected = true;
     }
 
-    QString platformThemeName = QString::fromLocal8Bit(qgetenv("QT_QPA_PLATFORMTHEME"));
+    platformThemeOverride = QString::fromLocal8Bit(qgetenv("QT_QPA_PLATFORMTHEME"));
 
     // Get command line params
 
@@ -1507,7 +1494,7 @@ void QGuiApplicationPrivate::createPlatformIntegration()
             }
         } else if (strcmp(arg, "-platformtheme") == 0) {
             if (++i < argc)
-                platformThemeName = QString::fromLocal8Bit(argv[i]);
+                platformThemeOverride = QString::fromLocal8Bit(argv[i]);
         } else if (strcmp(arg, "-qwindowgeometry") == 0 || (xcbIsDefault && strcmp(arg, "-geometry") == 0)) {
             if (++i < argc)
                 windowGeometrySpecification = QWindowGeometrySpecification::fromArgument(argv[i]);
@@ -1530,7 +1517,7 @@ void QGuiApplicationPrivate::createPlatformIntegration()
 
     Q_UNUSED(platformExplicitlySelected);
 
-    init_platform(QLatin1StringView(platformName), platformPluginPath, platformThemeName, argc, argv);
+    init_platform(QLatin1StringView(platformName), platformPluginPath, argc, argv);
     if (const QPlatformTheme *theme = platformTheme())
         QStyleHintsPrivate::get(QGuiApplication::styleHints())->setColorScheme(theme->colorScheme());
 
diff --git a/src/gui/kernel/qguiapplication_p.h b/src/gui/kernel/qguiapplication_p.h
index 58c3f33394..30309af72a 100644
--- a/src/gui/kernel/qguiapplication_p.h
+++ b/src/gui/kernel/qguiapplication_p.h
@@ -267,6 +267,7 @@ public:
 
     static QFont *app_font;
 
+    static QString platformThemeOverride;
     static QString styleOverride;
     static QStyleHints *styleHints;
     static bool obey_desktop_settings;
diff --git a/src/gui/platform/unix/qgenericunixthemes.cpp b/src/gui/platform/unix/qgenericunixthemes.cpp
index ce729a74a3..4703bb799c 100644
--- a/src/gui/platform/unix/qgenericunixthemes.cpp
+++ b/src/gui/platform/unix/qgenericunixthemes.cpp
@@ -1314,24 +1314,13 @@ QStringList QGenericUnixTheme::themeNames()
     QStringList result;
     if (QGuiApplication::desktopSettingsAware()) {
         const QByteArray desktopEnvironment = QGuiApplicationPrivate::platformIntegration()->services()->desktopEnvironment();
-        QList<QByteArray> gtkBasedEnvironments;
-        gtkBasedEnvironments << "GNOME"
-                             << "X-CINNAMON"
-                             << "UNITY"
-                             << "MATE"
-                             << "XFCE"
-                             << "LXDE";
         const QList<QByteArray> desktopNames = desktopEnvironment.split(':');
+        result.push_back(QStringLiteral("xdgdesktopportal"));
         for (const QByteArray &desktopName : desktopNames) {
             if (desktopEnvironment == "KDE") {
 #if QT_CONFIG(settings)
                 result.push_back(QLatin1StringView(QKdeTheme::name));
 #endif
-            } else if (gtkBasedEnvironments.contains(desktopName)) {
-                // prefer the GTK3 theme implementation with native dialogs etc.
-                result.push_back(QStringLiteral("gtk3"));
-                // fallback to the generic Gnome theme if loading the GTK3 theme fails
-                result.push_back(QLatin1StringView(QGnomeTheme::name));
             } else {
                 // unknown, but lowercase the name (our standard practice) and
                 // remove any "x-" prefix
@@ -1339,6 +1328,10 @@ QStringList QGenericUnixTheme::themeNames()
                 result.push_back(s.startsWith("x-"_L1) ? s.mid(2) : s);
             }
         }
+        // prefer the GTK3 theme implementation with native dialogs etc.
+        result.push_back(QStringLiteral("gtk3"));
+        // fallback to the generic Gnome theme if loading the GTK3 theme fails
+        result.push_back(QLatin1StringView(QGnomeTheme::name));
     } // desktopSettingsAware
     result.append(QLatin1StringView(QGenericUnixTheme::name));
     return result;
diff --git a/src/plugins/platformthemes/xdgdesktopportal/qxdgdesktopportaltheme.cpp b/src/plugins/platformthemes/xdgdesktopportal/qxdgdesktopportaltheme.cpp
index 355d3e6cc9..4640e83463 100644
--- a/src/plugins/platformthemes/xdgdesktopportal/qxdgdesktopportaltheme.cpp
+++ b/src/plugins/platformthemes/xdgdesktopportal/qxdgdesktopportaltheme.cpp
@@ -85,26 +85,26 @@ QXdgDesktopPortalTheme::QXdgDesktopPortalTheme()
     Q_D(QXdgDesktopPortalTheme);
 
     QStringList themeNames;
+    if (!QGuiApplicationPrivate::platformThemeOverride.isEmpty())
+        themeNames += QGuiApplicationPrivate::platformThemeOverride;
     themeNames += QGuiApplicationPrivate::platform_integration->themeNames();
-    // 1) Look for a theme plugin.
+    themeNames.removeAll(QStringLiteral("xdgdesktopportal"));
+    themeNames.removeAll(QStringLiteral("flatpak"));
+    themeNames.removeAll(QStringLiteral("snap"));
+
+    // 1) Look for a theme.
     for (const QString &themeName : std::as_const(themeNames)) {
         d->baseTheme = QPlatformThemeFactory::create(themeName, nullptr);
         if (d->baseTheme)
             break;
-    }
 
-    // 2) If no theme plugin was found ask the platform integration to
-    // create a theme
-    if (!d->baseTheme) {
-        for (const QString &themeName : std::as_const(themeNames)) {
-            d->baseTheme = QGuiApplicationPrivate::platform_integration->createPlatformTheme(themeName);
-            if (d->baseTheme)
-                break;
-        }
-        // No error message; not having a theme plugin is allowed.
+        d->baseTheme = QGuiApplicationPrivate::platform_integration->createPlatformTheme(themeName);
+        if (d->baseTheme)
+            break;
     }
+    // No error message; not having a theme is allowed.
 
-    // 3) Fall back on the built-in "null" platform theme.
+    // 2) Fall back on the built-in "null" platform theme.
     if (!d->baseTheme)
         d->baseTheme = new QPlatformTheme;
 
